### PR TITLE
Tolerate GPU taint

### PR DESCRIPTION
**Pull request by Furisto**
## Description
Ensures that ws-daemon and registry-facade can tolerate taints on GPU nodes.

## Related Issue(s)

Fixes ENT-201

#### Preview status

gitpod:summary

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [x] /werft preemptible
      Saves cost. Untick this only if you're really sure you need a non-preemtible machine.
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: install/installer/pkg/components/registry-facade/daemonset.go [hunk 1]
```diff
@@ -189,6 +189,13 @@ func daemonset(ctx *common.RenderContext) ([]runtime.Object, error) {
 					RestartPolicy:                 corev1.RestartPolicyAlways,
 					TerminationGracePeriodSeconds: pointer.Int64(30),
 					InitContainers:                initContainers,
+					Tolerations: []corev1.Toleration{
+						{
+							Key:      "gitpod.io/gpu",
+							Operator: "Exists",
+							Effect:   "NoSchedule",
```
Comment (geropl) [inline install/installer/pkg/components/registry-facade/daemonset.go:196]: nit: To my understanding we do not need the `Effect` here with operator `Exists`

From the [docs](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/#concepts):
> The default value for operator is Equal.
> A toleration "matches" a taint if the keys are the same and the effects are the same, and:
> - the operator is Exists (in which case no value should be specified), or
> - the operator is Equal and the values should be equal.

File: install/installer/pkg/components/ws-daemon/daemonset.go [hunk 2]
```diff
@@ -204,6 +204,11 @@ func daemonset(ctx *common.RenderContext) ([]runtime.Object, error) {
 			Operator: "Exists",
 			Effect:   "NoExecute",
 		},
+		{
+			Key:      "gitpod.io/gpu",
+			Operator: "Exists",
+			Effect:   "NoSchedule",
```
Comment (geropl) [inline install/installer/pkg/components/ws-daemon/daemonset.go:210]: Same as [here](https://github.com/gitpod-io/gitpod/pull/19791/files#r1618714523).

**Review by geropl — APPROVED** [review]
Changes LGTM :heavy_check_mark: 
Comments are nit only

File: install/installer/pkg/components/registry-facade/daemonset.go (hunk 1 above)
Comment (Furisto) [reply install/installer/pkg/components/registry-facade/daemonset.go:196]: @geropl 
I do not understand it this way. Just below of what you quoted (We only want to match `NoSchedule`)

> An empty effect matches all effects with key key1.

Comment (geropl) [reply install/installer/pkg/components/registry-facade/daemonset.go:196]: >  An empty effect matches all effects with key key1.

Ja, that section is badly worded. I understood that this line only applies to `Equal`, but who knows :sweat_smile: 

> (We only want to match NoSchedule)

:bulb: Ah, did not know! Then it is the right thing I guess. :+1:

**Pull request merged**